Add supports for PlantUML and WaveDrom diagrams

zhang-stephen opened the pull request:
This PR is opened for adding PlantUML and WaveDrom supports into DoIt theme of hugo. And I converted all CRLFs to LFs before my works, to finish the chaos of EOLs.

All of them are implemented by render hooks of codeblocks. But there are few difference between PlantUML and WaveDroms.

---

For PlantUML diagrams, all code blocks with type `plantuml` would be rendered as a `<img>` tag into HTML.

Codeblock for example:
~~~markdown
```plantuml {format="svg", title="example"}
[content]
```
~~~

It would be rendered as:

```html
<div id="plantuml-diagram">
     <img src="[server]/[format]/[encoded]" alt=[title]"/>
</div>
```

The explaining of `[]` parameters of the rendered tag `<img>`:

param | description | type
--- | --- | ---
`server` | the URL of online rendering server, which could be configured by `params.page.plantuml.server`, the default is https://www.plantuml.com/plantuml | string
`format` | the output format of PlantUML renderer, could be one of `svg` (default) or `png`, while the `txt` for AsciiDoc is not supported | string 
`title` | would be rendered as the attribute `alt` of tag `<img>`, and it's optional | string
`encoded` | the encoded PlantUML code | string

The `<img>` would be aligned in the center of its container, and would be zoomed with the maximum width of content.

The dependency is [plantuml-encoder](https://github.com/markushedvall/plantuml-encoder), an offically recommanded encoder implementation with javascript.

---

For the WaveDrom, it's a powerful toolkit for IC designers, embeded software engineers (like me) and other related works. All codeblocks with type `wavedrom` would be rendered as embeded SVG into HTML by 2-phased rendering.

Codeblock for example:

~~~markdown
```wavedrom
{signal: [
  {name: 'clock',  wave: 'p...'},
  {name: 'A1',     wave: ['pw', {d: ['M',0,0, 'L',1,1, 'L',2,1, 'L',3,0, 'H',4]}]}
]}
```
~~~

The 1st phase of rendering (render hook):

```html
<script type="WaveDrom">
{signal: [
  {name: 'clock',  wave: 'p...'},
  {name: 'A1',     wave: ['pw', {d: ['M',0,0, 'L',1,1, 'L',2,1, 'L',3,0, 'H',4]}]}
]}
</script>
```

The 2nd phase of rendering (WaveDorm.js), while the content of `<svg>` would be omitted:

```html
<div id="WaveDrom_Display_[ID]">
     <svg>...</svg>
</div>
```

The rendered SVG would be aligned to the left for long diagrams.

Due to the limitation of WaveDrom itself, the default skin would be supported only. And there is no extra options to be configured.

The dependency is [wavedrom](https://github.com/wavedrom/wavedrom).

---

This PR is finished with some help from LLM (deepseek-R1). Please help to review the PR and merge it, thanks.

zhang-stephen committed:
feat: add support for PlantUML diagrams

zhang-stephen committed:
feat: add support for WaveDrom diagrams

HEIGE-PCloud reviewed (commented):
Can you cherry pick the  CRLF to LF conversion into a separate PR and we can merge that one first so the diff is more readable?

zhang-stephen commented:
> Can you cherry pick the CRLF to LF conversion into a separate PR and we can merge that one first so the diff is more readable?

PR #1458 has been created and ready for review. And should I recreate a PR for this after it merged?

HEIGE-PCloud commented:
> PR #1458 has been created and ready for review. And should I recreate a PR for this after it merged?


#1458 is merged. I merged main into this so there is no need to recreate the PR.

HEIGE-PCloud commented on assets/js/theme.ts:
```diff
@@ -928,6 +928,48 @@ function initCodeblocks() {
+    renderPlantUML();
+    renderWaveDrom();
+  });
+}
```
Wait on `DOMContentLoaded` is unnecessary, `theme.ts` is marked as `defer`.

HEIGE-PCloud commented on assets/js/theme.ts:
```diff
@@ -928,6 +928,48 @@ function initCodeblocks() {
+      diagram.style.color = "red";
+    }
+  });
+}
```
Split initialisation into a separate file, place it here: https://github.com/HEIGE-PCloud/DoIt/tree/faff264ea1bdd8bf87a50f66cc6637192c8e32bf/assets/js/lib and also load it conditionally.

We don't need to load the initialisation code unless PlantUML is enabled on the page.

zhang-stephen replied:
It has been moved into `assets/js/lib/plantuml.js` with conditional loading.

HEIGE-PCloud commented on assets/js/theme.ts:
```diff
@@ -928,6 +928,48 @@ function initCodeblocks() {
+
+function renderWaveDrom() {
+  WaveDrom.ProcessAll();
+}
```
Same for WaveDrom

zhang-stephen replied:
It has been moved into assets/js/lib/wavedrom.js with conditional loading.

HEIGE-PCloud commented on assets/js/theme.ts:
```diff
@@ -928,6 +928,48 @@ function initCodeblocks() {
+      console.error(errMsg);
+      diagram.innerHTML = errMsg;
+      diagram.style.color = "red";
+    }
```
Does it ever throw?

zhang-stephen replied:
Yes, it would be throwed if the `${server}` is invalid. I think this error message could be kept, do you have more better advice for it?

HEIGE-PCloud replied:
Sorry I don't quite understand the argument here, are you talking about line 944?

I think `img` with invalid `src` will just show a broken image. I mean if there is no exception, there is no need to try catch.

zhang-stephen replied:
Yes, I am talking about the line 944. And this catch block has been removed.

HEIGE-PCloud commented on assets/css/_page/_single.scss:
```diff
@@ -336,3 +336,19 @@
+  border-radius: 0.25rem;
+  justify-content: center;
+  align-items: center;
+}
```
This is unnecessary, just set the image with margin auto.

zhang-stephen replied:
done.

HEIGE-PCloud commented on assets/css/_page/_single.scss:
```diff
@@ -336,3 +336,19 @@
+  // justify-content: center; // aligned to left for long diagrams
+  align-items: center;
+  background-color: #FCFCFA;
+}
```
This is also unnecessary, just do

```css
[id^=WaveDrom_Display] > svg {
  max-width: 100%;
  margin: 0 auto;
}
```

zhang-stephen replied:
Done. But the `background-color` has been kept for Wavedrom diagrams with transparent background.

zhang-stephen commented:
> > PR #1458 has been created and ready for review. And should I recreate a PR for this after it merged?
> 
> #1458 is merged. I merged main into this so there is no need to recreate the PR.

Hi, would you accept this PR? 

I found it is too hard to maintain supports of PlantUML and WaveDrom in my own blog repo, it causes too much duplicated code in `layouts/` of the blog root, due to the template look-up rules of hugo. Then I think the better solution might be merging them into mainline.

HEIGE-PCloud commented on layouts/_default/_markup/render-codeblock-plantuml.html:
```diff
@@ -0,0 +1,18 @@
+        data-content="{{ .Inner | safeHTML }}"
+        data-server="{{ $server }}"
+        data-format="{{ $format }}"
+    />
```
`pre` is a non-void HTML element. It needs to be closed by `<pre>data</pre>`.

Why use `pre` here? An empty `pre` tag has impact on the layout. Can we use a `script` tag just like WaveDrom?

zhang-stephen replied:
Done. It has been replaced by a `<script type='plantuml'>`.

HEIGE-PCloud commented on exampleSite/content/posts/theme-documentation-content/index.zh-cn.md:
For consistency, we add a space between Chinese and English characters https://github.com/sparanoid/chinese-copywriting-guidelines?tab=readme-ov-file#中英文之間需要增加空格

zhang-stephen replied:
done.

HEIGE-PCloud reviewed (commented):
I would suggest adding tests in https://github.com/HEIGE-PCloud/DoIt/tree/a7af0778f1256747fa1d92c1f1d8b79b27850438/exampleSite/content/posts/tests and add visual regression test cases for them here https://github.com/HEIGE-PCloud/DoIt/blob/a7af0778f1256747fa1d92c1f1d8b79b27850438/tests/visual.spec.ts

HEIGE-PCloud commented:
> Hi, would you accept this PR?
> 
> I found it is too hard to maintain supports of PlantUML and WaveDrom in my own blog repo, it causes too much duplicated code in `layouts/` of the blog root, due to the template look-up rules of hugo. Then I think the better solution might be merging them into mainline.

This PR is mostly fine, I am happy to accept it. I mean DoIt supports Mermaid https://hugodoit.pages.dev/create-diagrams/#mermaid and this is similar. But is this the best way to embed these diagrams? In my opinion it's quite wasteful to embed the source in the html, load a client side javascript file just to call a remote API endpoint to render a png image for every single visitor of your site every time. Why not editing these diagrams in an editor locally and export them as png/svg files and directly embed these images in your site?

zhang-stephen commented:
> > Hi, would you accept this PR?
> > I found it is too hard to maintain supports of PlantUML and WaveDrom in my own blog repo, it causes too much duplicated code in `layouts/` of the blog root, due to the template look-up rules of hugo. Then I think the better solution might be merging them into mainline.
> 
> This PR is mostly fine, I am happy to accept it. I mean DoIt supports Mermaid https://hugodoit.pages.dev/create-diagrams/#mermaid and this is similar. But is this the best way to embed these diagrams? In my opinion it's quite wasteful to embed the source in the html, load a client side javascript file just to call a remote API endpoint to render a png image for every single visitor of your site every time. Why not editing these diagrams in an editor locally and export them as png/svg files and directly embed these images in your site?

Thanks for your comments. I will refactor this PR and resolve the comments in next few days. 

And there are few reasons why I need this PR:

- For PlantUML:
  - Mermaid has its own advantages, e.g. git graph, mindmap or pie graph, but it cannot replace PlantUML at all for sequence diagram or class diagram. And for some historical reasons, I have to use PlantUML in these scenarios for daily works.
  - And the PlantUML is implemented with Java only (Its some components could be replaced by VizJS but works not well), so it must be rendered by calling remote API.
  - I perfer the SVG format for PlantUML, the `png` is supported by the way.
- For WaveDrom, it would be rendered like the Mermaid. Even the bit-field diagram or something like is supported by Mermaid, but WaveDrom is more powerful than Mermaid. Besides, WaveDrom supports more diagrams for embedded software engineering.
- For myself, I prefer online rendering diagrams than pre-rendered diagrams:
  - I deployed a hugo blog in work intranet, then I cannot upload images or diagrams to public image hosting service.
  - This blog is deployed to intranet gitlab instance, but there is no better way to manage images by git (git LFS is disabled for personal repo).
  - It is hard to update pre-rendered diagrams if the source file lost.

zhang-stephen committed:
chore: add spaces between Chinese and English words

zhang-stephen committed:
fix: remove the rendering for PlantUML & Wavedrom from theme.js

zhang-stephen committed:
refactor: move the plantuml rendering function into js/lib
- use <script> tag to load the plantuml code
- add default id for renderer <img> for CSS selector
- simplify the plantuml style sheet

zhang-stephen committed:
refactor: move the rendering function into js/lib
- simplify the stylesheet of wavedrom diagrams

zhang-stephen committed:
chore: format code with prettier

zhang-stephen committed:
feat: add regression tests for rendering of PlantUML and WaveDrom

zhang-stephen committed:
revert: refactor: move the rendering function into js/lib

- revert changes for _single.scss
- the others would be re-applied in the next commit

This reverts commit 0b25a389932a3c3137e1a69f8628aa24ef3f0018.

zhang-stephen committed:
refactor: move the wavedrom rendering into js/lib for conditional loading

zhang-stephen commented:
Hi, please review the new changes again, and kindly help to fix the failure of playwright regression failed.

zhang-stephen committed:
fix: reset the code format which is broken by the vscode formatter

- remove unexpected yarn lock

HEIGE-PCloud committed:
refactor: remove unnecessary min-height class from main.css

HEIGE-PCloud committed:
feat: support light dark theme switch for plantuml wavedrom

HEIGE-PCloud committed:
test: add visual regression test for plantuml page

HEIGE-PCloud committed:
refactor: remove plantuml-tests from visual test suite

HEIGE-PCloud committed:
fix: correct BASE_URL usage in visual regression tests

HEIGE-PCloud committed:
test: update visual regression test to target specific plantuml diagrams

HEIGE-PCloud committed:
test: enhance visual regression tests by forcing lazy-loaded images to load eagerly

HEIGE-PCloud committed:
test: mark visual regression test for plantuml as a fixme
